src/node/tendermint.ml

```diff
@@ -0,0 +1,206 @@
+open Helpers
+open Tendermint_internals
+open Tendermint_data
+open Tendermint_processes
+open Tendermint_helpers
+
+type t = {
+  identity : State.identity;
+  clocks : clock list IntSet.t;
+  consensus_states : consensus_state IntSet.t;
+  (* TODO: make non-mutable *)
+  mutable procs : (height * process) list;
+  node_state : State.t;
+  input_log : input_log;
+  output_log : OutputLog.t;
+}
+
+type should_restart_tendermint =
+  | DontRestart
+  | RestartAtHeight of height
+  | RestartAtRound  of round
+
+let make node_state current_height =
+  let identity = node_state.State.identity in
+  let clocks = IntSet.create 0 in
+  let new_state = fresh_state current_height in
+  let states = IntSet.create 0 in
+  IntSet.add states current_height new_state;
+  let procs = List.map (fun x -> (current_height, x)) all_processes in
+  let input_log = empty () in
+  let output_log = OutputLog.empty () in
+  {
+    identity;
+    clocks;
+    consensus_states = states;
+    node_state;
+    procs;
+    input_log;
+    output_log;
+  }
+
+let current_height node =
+  List.fold_left max 0L (IntSet.to_seq_keys node.consensus_states |> List.of_seq)
+
+(* FIXME: bad design *)
+let () =
+  produce_value := fun state -> block (Building_blocks.produce_block state)
+
+(** Process messages in the queue and decide actions; pure function, to be interpeted in Lwt later.
+    TODO: Ensures that messages received over network go through signature verification before adding them to input_log
+    FIXME: we're not empyting the input_log atm *)
+let tendermint_step node =
+  let open Tendermint_processes in
+  let rec exec_procs processes still_active network_actions =
+    match processes with
+    | ((height, process) as p) :: rest -> begin
+      let consensus_state = IntSet.find node.consensus_states height in
+      let round = consensus_state.Tendermint_internals.round in
+      let message_log = node.input_log in
+      let output = node.output_log in
+      match
+        process height round consensus_state message_log output node.node_state
+      with
+      (* The process precondition hasn't been activated *)
+      | None -> exec_procs rest (p :: still_active) network_actions
+      (* The process terminates with a network event *)
+      | Some (Broadcast t) -> exec_procs rest still_active (t :: network_actions)
+      (* The process terminates silently *)
+      | Some DoNothing -> exec_procs rest still_active network_actions
+      (* We accepted a block for the height *)
+      | Some (RestartTendermint (height, 0)) ->
+        (* Start new processes and forget about the older ones *)
+        (* TODO: this is only valid for the (current) restricted version of Tendermint which does not
+           support several cycles/heights running in parallel. *)
+        ([], [], RestartAtHeight height)
+      | Some (RestartTendermint (_height, round)) ->
```

Comment:
Where is this `_height` coming from?

Reply:
In `tendermint_processes.ml` we have:
```
type consensus_action =
  | Broadcast         of sidechain_consensus_op
  | Schedule          of consensus_action option Clock.t
  | RestartTendermint of (height * round)
  | DoNothing
  ```
  
  Here, we discard `_height`